### PR TITLE
Use Fabric's quantization

### PR DESCRIPTION
**Pull request by carmocca**
Fixes https://github.com/Lightning-AI/lit-gpt/issues/425

### CODE REVIEW

File: tutorials/finetune_lora.md [hunk 1]
```diff
@@ -59,25 +59,21 @@ python finetune/lora.py --quantize "bnb.nf4-dq"
...
+| Settings                                    | Training Memory | Training Time | Loss   | Inference Memory |
+|---------------------------------------------|-----------------|---------------|--------|------------------|
+| Default (bf16-mixed)                        | 38.01 GB        | 591.78s       | 0.9207 | 21.49 GB         |
+| --precision bf16-true                       | 19.40 GB        | 592.14s       | 0.9180 | 7.35 GB          |
+| --precision bf16-true --quantize bnb.nf4    | 19.13 GB        | 802.02s       | 0.9408 | 3.26 GB          |
+| --precision bf16-true --quantize bnb.nf4-dq | 18.97 GB        | 802.94s       | 0.9384 | 3.09 GB          |
 
 The advantages of QLoRA-style quantization are more pronounced in larger models, such as Llama 2 7B. The table below summarizes the results for Llama 2 7B on Alpaca for 5,000 iterations using a microbatch size of 4:
 
-| Settings                                            | Training Memory  | Training Time  | Loss   | Inference Memory |
```
Comment (carmocca) [inline tutorials/finetune_lora.md:73]: @rasbt I'm unable to reproduce your original training memory results after checking out https://github.com/Lightning-AI/lit-gpt/pull/275. Can you help?

File: tutorials/finetune_lora.md [hunk 2]
```diff
@@ -59,25 +59,21 @@ python finetune/lora.py --quantize "bnb.nf4-dq"
 
 The table below lists a comparison with different settings on a StableLM 3B model finetuned with LoRA on Alpaca for 5,000 iterations using a microbatch size of 4:
 
-| Settings                                                | Training Memory  | Training Time  | Loss      | Inference Memory |
```
Comment (carmocca) [inline tutorials/finetune_lora.md:62]: I haven't been able to reproduce these memory values even after checking the original PR that added them, but this PR's values are consistent with current main

File: lit_gpt/utils.py [hunk 3]
```diff
@@ -25,62 +26,30 @@ def find_multiple(n: int, k: int) -> int:
...
-
-        class QuantizedLinear(bnb.Linear4bit):
-            def __init__(self, *args, **kwargs):
-                super().__init__(*args, quant_type="nf4", compress_statistics=False, **kwargs)
+    total = 0
+    for p in module.parameters():
+        if requires_grad is None or p.requires_grad == requires_grad:
+            if hasattr(p, "quant_state"):
+                # bitsandbytes 4bit layer support
+                total += math.prod(p.quant_state[1])
```
Comment (carmocca) [inline lit_gpt/utils.py:34]: The count is off otherwise with quantized linears

File: lit_gpt/speed_monitor.py [hunk 4]
```diff
@@ -293,10 +294,28 @@ def eval_end(self, eval_elapsed: float):
...
+    if isinstance(plugin, TransformerEnginePrecision):
+        return torch.int8
+    if isinstance(plugin, (FSDPPrecision, FSDPPrecisionPlugin)):
+        return plugin.mixed_precision_config.reduce_dtype
+    raise NotImplementedError(plugin)
+
+
 class SpeedMonitorFabric(SpeedMonitorBase):
     def __init__(self, fabric: Fabric, *args: Any, **kwargs: Any) -> None:
-        # TODO: this will not work properly if a precision plugin is passed to Fabric
```
Comment (carmocca) [inline lit_gpt/speed_monitor.py:298]: The changes in this file are meant to resolve this TODO

File: lit_gpt/speed_monitor.py [hunk 5]
```diff
@@ -293,10 +294,28 @@ def eval_end(self, eval_elapsed: float):
         self.total_eval_wct += eval_elapsed  # seconds
 
 
+def plugin_to_compute_dtype(plugin: Precision) -> torch.dtype:
```
Comment (carmocca) [inline lit_gpt/speed_monitor.py:297]: This could be in Fabric/PL

**Comment by carmocca** [comment]
This is ready to be merged after the linked PRs are merged upstream

File: chat/base.py [hunk 6]
```diff
@@ -135,24 +141,32 @@ def main(
     """
     precision = precision or get_default_supported_precision(training=False)
 
+    plugins = None
+    if quantize is not None and quantize.startswith("bnb."):
+        if "mixed" in precision:
+            raise ValueError("Quantization and mixed precision is not supported.")
```
Comment (awaelchli) [inline chat/base.py:147]: I think we should eventually check this in Fabric's connector directly (no blocker ofc).

File: requirements.txt [hunk 7]
```diff
@@ -5,8 +5,6 @@ jsonargparse[signatures]  # CLI
 # other optional dependencies are
 # sentencepiece  # pythia, falcon, redpajama
 # tokenizers  #  llama-based models
-# bitsandbytes>=0.41.1  # quantize/bnb.py
```
Comment (awaelchli) [inline requirements.txt:8]: This is still an optional dependency that you need for quantization, so shouldn't it stay?

File: lit_gpt/speed_monitor.py (hunk 5 above)
Comment (awaelchli) [reply lit_gpt/speed_monitor.py:297]: Maybe something like `plugin.compute_dtype` and `plugin.param_dtype`.

File: lit_gpt/lora.py [hunk 8]
```diff
@@ -120,8 +120,8 @@ def __init__(
 
         # Actual trainable parameters
         if r > 0:
-            self.lora_A = nn.Parameter(self.linear.weight.new_zeros((r, in_features)))
```
Comment (awaelchli) [inline lit_gpt/lora.py:123]: Are these changes needed? The previous way is not wrong, right?

File: generate/base.py [hunk 9]
```diff
@@ -137,22 +156,15 @@ def main(
...
-    with fabric.init_module(empty_init=True), quantization(quantize):
+    with fabric.init_module(empty_init=True), gptq_quantization(quantize == "gptq.int4"):
         model = GPT(config)
     fabric.print(f"Time to instantiate model: {time.perf_counter() - t0:.02f} seconds.", file=sys.stderr)
 
-    t0 = time.perf_counter()
-
-    if quantize:
-        # for quantization, need to load before moving to device
-        load_checkpoint(fabric, model, checkpoint_path, strict=(quantize is None))
```
Comment (awaelchli) [inline generate/base.py:148]: I had to put it here because it would fail for some modes. Does it work for all modes now?

File: generate/adapter.py [hunk 10]
```diff
@@ -77,15 +93,15 @@ def main(
...
-    with fabric.init_module(empty_init=True), quantization(quantize):
+    with fabric.init_module(empty_init=True), gptq_quantization(quantize == "gptq.int4"):
         model = GPT(config)
     fabric.print(f"Time to instantiate model: {time.perf_counter() - t0:.02f} seconds.", file=sys.stderr)
 
     t0 = time.perf_counter()
     checkpoint = lazy_load(checkpoint_path)
     adapter_checkpoint = lazy_load(adapter_path)
     checkpoint.update(adapter_checkpoint.get("model", adapter_checkpoint))
-    model.load_state_dict(checkpoint, strict=quantize is None)
```
Comment (awaelchli) [inline generate/adapter.py:88]: Is this resolved by the new state dict hooks we added?

File: eval/lm_eval_harness.py [hunk 11]
```diff
@@ -1,12 +1,11 @@
 import json
```
Comment (awaelchli) [inline eval/lm_eval_harness.py:1]: I suggest to ask the harness script author for review.

File: tutorials/finetune_lora.md (hunk 2 above)
Comment (awaelchli) [reply tutorials/finetune_lora.md:62]: Did you also check whether peak memory usage is consistent with main? Because that would be important regarding instantiation + quantization.

**Review by awaelchli — APPROVED** [review]
Code changes LGTM 🎉 , but I didn't run any comparisions with main. Let me know if you need help running experiments if not already done.

File: chat/base.py (hunk 6 above)
Comment (carmocca) [reply chat/base.py:147]: https://github.com/Lightning-AI/lightning/pull/18671 would raise an error already

File: requirements.txt (hunk 7 above)
Comment (carmocca) [reply requirements.txt:8]: note that this is commented out, the tutorials for quantization still mention that it should be installed.

The pin is now decided by lightning, not by lit-gpt. This is why I removed it

File: lit_gpt/lora.py (hunk 8 above)
Comment (carmocca) [reply lit_gpt/lora.py:123]: They fix https://github.com/Lightning-AI/lit-gpt/issues/425

File: generate/adapter.py (hunk 10 above)
Comment (carmocca) [reply generate/adapter.py:88]: Yes, exactyl

File: tutorials/finetune_lora.md (hunk 2 above)
Comment (carmocca) [reply tutorials/finetune_lora.md:62]: That is what I meant with

> this PR's values are consistent with current main
